packages/runtime/src/state/actions.ts

```diff
@@ -378,12 +378,21 @@ export function cleanUp(): CleanUpAction {
   return { type: ActionTypes.CLEAN_UP }
 }
 
-export function registerDocument(document: Document): RegisterDocumentAction {
-  return { type: ActionTypes.REGISTER_DOCUMENT, payload: { documentKey: document.key, document } }
+export function registerDocument(
+  document: Document,
+  builderInitiated = false,
+): RegisterDocumentAction {
```

Comment:
With the introduction of the middleware, initiating a `registerDocument` via the builder will result in a ping-pong. As of now, this only applies to the case of focusing a global element:

1. Builder sends a `registerDocument` action via post message
2. Host receives action, registers a document. This will also kick off the middleware again, sending the document registration back to the builder.
3. When the builder receives the registration, the builder attempts to resubscribe to things since the document keys it has received have changed. This results in other actions getting missed since they are improperly buffered, preventing the global element from being interactive.

To resolve this, the builder can now call this method and pass `builderInitiated = true` to prevent the ping-ponging. Previous versions of the runtime won't have a middleware, so the ping-ponging won't happen anyways. The payload itself is backwards compatible since it's a new key.

I imagine this particular portion of the PR will be the area of most contention, so it's worth noting that I was able to get things working in the builder **without** having this carve out. However, it temporarily involves some tricks with our ShareDB reducers, and long-term (i.e., maybe this cooldown?) should use some refactoring that I've been wanting to do anyways. 

No matter what, some builder change will be required: right now, shipping the middleware alone will result in race conditions when attempting to edit global elements.

Reply:
I'm sure I'm missing something here, but since we are effectively translating `REGISTER_DOCUMENT` to a `registerBuilderDocument` call, can builder just dispatch `REGISTER_BUILDER_DOCUMENT` instead?

Reply:
Not quite, since the builder dispatching `REGISTER_DOCUMENT` serves the purpose of populating the runtime store, which in turn changes what the host renders. `REGISTER_BUILDER_DOCUMENT` is solely for the host to communicate the mounting of a document to the builder, so dispatching that directly from the builder would serve no purpose.

Reply:
This is an unfortunate piece of technical debt that stems from using this action to "register" a document when a global component is focused. This is far from ideal. Instead, an action to "isolate" or "focus" a document should allow the runtime to copy the relevant document and then scope received ops to that. That might be a better solution. I'm sure there's also others that are far simpler than the status quo.